src/antsibull/build_ansible_commands.py

```diff
@@ -345,13 +348,32 @@ def _is_alpha(version: PypiVer) -> bool:
     return version.is_prerelease and pre is not None and pre[0] == 'a'
 
 
+def _extract_python_requires(ansible_core_version: PypiVer, deps: t.Dict[str, str]) -> str:
+    python_requires = deps.pop('_python', None)
+    if python_requires is not None:
+        return python_requires
+    if ansible_core_version < PypiVer('2.12.0a'):
+        # Ansible 2.9, ansible-base 2.10, and ansible-core 2.11 support Python 2.7 and Python 3.5+
+        return '>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*'
+    if ansible_core_version < PypiVer('2.14.0a'):
+        # ansible-core 2.12 and 2.13 support Python 3.8+
+        return '>=3.8'
+    if ansible_core_version < PypiVer('2.15.0a'):
+        # ansible-core 2.14 supports Python 3.9+
+        return '>=3.9'
+    raise ValueError(
+        f'Python requirements for ansible-core {ansible_core_version} should be part of'
```

Comment:
I don't understand why you think setting a `python_requires` is even necessary in the `ansible` package. pip already inspects `python_requires` of dependencies like `ansible-core` as a part of its dependency resolution process and acts accordingly.

Reply:
Do old versions of pip (that handle `python_requires` at all) also do that?

Anyway, I think this information is useful even if it can be indirectly derived, since https://pypi.org/project/ansible/ also shows that information (in `Meta`), and I'm not sure whether it would do that when the package wouldn't explicitly declare it. Also it makes this value available in https://pypi.org/pypi/ansible/json, which other programs might find useful.

Reply:
I disagree. Ansible is tightly coupled to ansible-core, so it exposing the same `python_requires` as ansible-core is not a premature optimization.

> OTOH, since `ansible` bundles the code from collections, it could derive its value from them and specify whatever the collections declare (well, the common denominator, I guess), since those collections are the actual content of the distribution being shipped.

That would be totally wrong from my POV. Just because one collection requires Python 3.10 it doesn't mean that other collections cannot work with Python 3.9. Besides collections cannot declare their supported Pyton version, so this discussion is moot anyway.

Reply:
> I disagree. Ansible is tightly coupled to ansible-core, so it exposing the same `python_requires` as ansible-core is not a premature optimization.

Coupled, yes. But does not bundle it. Which is why it has its own metadata (name, version, deps). And `Requires-Python` is just one of those dependency declarations.

> > OTOH, since `ansible` bundles the code from collections, it could derive its value from them and specify whatever the collections declare (well, the common denominator, I guess), since those collections are the actual content of the distribution being shipped.
> 
> That would be totally wrong from my POV. Just because one collection requires Python 3.10 it doesn't mean that other collections cannot work with Python 3.9. Besides collections cannot declare their supported Pyton version, so this discussion is moot anyway.

Okay, the right solution would be the lowest supported by any collection, then.

Reply:
I agree with @felixfontein here. The only reasonable `python_requires` for ansible is the same one as ansible-core. It wouldn't make sense to include Python 2.7 in the version range even though most modules still support that. Modules aren't run by the ansible controller. The requirement for controller plugins is that they support and interface with the ansible bundle's ansible-core version. Sure, in isolation, many controller plugins will work with older ansible-core/Python versions, but that's not relevant to the ansible bundle. *Controller* plugins in ansible 6 by definition aren't compatible with Python 3.6, because *the controller*, ansible-core 2.13, doesn't support that. Setting this to the same value as ansible-core is the most reasonable approach, and it provides a stronger defence against dependency issues, even if pip is supposed to figure this out on its own.

Reply:
I never talked about modules that execute on remote targets, though. You're right, only what executes on the controller should influence the metadata.
But, with my PyPA hat on, my argument is that this is a premature attempt to solve a non-existing problem by declaring unnecessary metadata. If you think that this theoretical problem with ansible-core is actually possible, would you mind demonstrating it?

Reply:
The controller-side plugins in community.general still support Python 2.7. (The libraries that some of these require might not, though.)

I still don't see why we should remove this value, even if it *would* be a premature optimization (which I disagree it is).

Reply:
I think that the argument about copying the restriction from ansible-core is weak because it's already there in the dependency. By doing so, we'd just be providing inaccurate metadata. But if the argument is that "we think that this bundle of collections should work in certain runtimes", maybe it'd be reasonable for as long as it's decided based on collections and factors that influence the project directly.

Reply:
> By doing so, we'd just be providing inaccurate metadata.

It is not inaccurate, it is very accurate. Ansible (the Python package) is ansible-core of a very specific version (which has a very specific Python requirement) combined with the content of collections.